### ISSUE TITLE
model.evaluate() does not yield the same accuracy as computing it manually using a for-loop

### ISSUE THREAD

**Issue by thor4** [tensorflow/tensorflow#50881]
Apologies in advance if this is the wrong place for this discussion. I suspect the `model.evaluate()` behavior I identify below is a feature, not a bug, but would like to better understand it if so.

After following the transfer learning tutorial on [Tensorflow's site][1], I have a question about how `model.evaluate()` works in comparison to calculating accuracy by hand.

At the very end, after fine-tuning, in the Evaluation and prediction section, we use `model.evaluate()` to calculate the accuracy on the test set as follows:
```
loss, accuracy = model.evaluate(test_dataset)
print('Test accuracy :', accuracy)
6/6 [==============================] - 2s 217ms/step - loss: 0.0516 - accuracy: 0.9740
Test accuracy : 0.9739583134651184
```

Next, we generate predictions manually from one batch of images from the test set as part of a visualization exercise:
```
# Apply a sigmoid since our model returns logits
predictions = tf.nn.sigmoid(predictions)
predictions = tf.where(predictions < 0.5, 0, 1)
```
However, it's also possible to extend this functionality to calculate predictions across the entire test set and compare them to the actual values to yield an average accuracy:
```
all_acc=tf.zeros([], tf.int32) #initialize array to hold all accuracy indicators (single element)
for image_batch, label_batch in test_dataset.as_numpy_iterator():
    predictions = model.predict_on_batch(image_batch).flatten() #run batch through model and return logits
    predictions = tf.nn.sigmoid(predictions) #apply sigmoid activation function to transform logits to [0,1]
    predictions = tf.where(predictions < 0.5, 0, 1) #round down or up accordingly since it's a binary classifier
    accuracy = tf.where(tf.equal(predictions,label_batch),1,0) #correct is 1 and incorrect is 0
    all_acc = tf.experimental.numpy.append(all_acc, accuracy)
all_acc = all_acc[1:]  #drop first placeholder element
avg_acc = tf.reduce_mean(tf.dtypes.cast(all_acc, tf.float16)) 
print('My Accuracy:', avg_acc.numpy()) 
My Accuracy: 0.974
```

Now, if `model.evaluate()` generates predictions by applying a sigmoid to the logit model outputs and using a threshold of 0.5 like the tutorial suggests, my manually-calculated accuracy should equal the accuracy output of Tensorflow's `model.evaluate()` function. This is indeed the case for the tutorial. My Accuracy: 0.974 = accuracy from `model.evaluate()` function. However, when I try this same code with a model trained using the same convolutional base as the tutorial, but different Gabor images (not cats & dogs like the tutorial), my accuracy no longer equals the `model.evaluate()` accuracy:
```
current_set = set17 #define set to process. 
all_acc=tf.zeros([], tf.float64) #initialize array to hold all accuracy indicators (single element)
loss, acc = model.evaluate(current_set) #now test the model's performance on the test set
for image_batch, label_batch in current_set.as_numpy_iterator():
    predictions = model.predict_on_batch(image_batch).flatten() #run batch through model and return logits
    predictions = tf.nn.sigmoid(predictions) #apply sigmoid activation function to transform logits to [0,1]
    predictions = tf.where(predictions < 0.5, 0, 1) #round down or up accordingly since it's a binary classifier
    accuracy = tf.where(tf.equal(predictions,label_batch),1,0) #correct is 1 and incorrect is 0
    all_acc = tf.experimental.numpy.append(all_acc, accuracy)
all_acc = all_acc[1:]  #drop first placeholder element
avg_acc = tf.reduce_mean(all_acc)
print('My Accuracy:', avg_acc.numpy()) 
print('Tf Accuracy:', acc) 
My Accuracy: 0.832
Tf Accuracy: 0.675000011920929
```
Does anyone know why there would be a discrepancy? Does the model.evaluate() *not* use a sigmoid? Or does it use a different threshold than 0.5? Or perhaps it's something else I'm not considering? Please note, my new model was trained using Gabor images, which are different than the cats and dogs from the tutorial, but the code was the same. 

Thank you in advance for any insight!

  [1]: https://www.tensorflow.org/tutorials/images/transfer_learning

**Comment by tilakrayal** [comment]
@thor4 ,

This question is better asked on [TensorFlow Forum](https://discuss.tensorflow.org/) since it is not a bug or feature request. There is also a larger community that reads questions there.

**Comment by thor4** [comment]
Thanks, I have posted it as you suggested: https://discuss.tensorflow.org/t/model-evaluate-does-not-yield-the-same-accuracy-as-computing-it-manually-using-a-for-loop/3214

Hopefully someone somewhere will have some insight.

**Comment by thor4** [comment]
I'm re-opening this again for your review as it may be a bug. There have been no responses on the Tensorflow Forum post and it's been a month now. Please advise.

**Comment by thor4** [comment]
thanks, I am working on a jupyter nodebook and will provide soon

**Comment by thor4** [comment]
Thanks for your patience. Here is a Jupyter notebook with the model and dataset which duplicates the accuracy calculation discrepancy: https://colab.research.google.com/github/thor4/neuralnets/blob/master/projects/1-CNN/model_evaluate.ipynb

Please let me know of questions/issues. Thank you for your help.

**Comment by tilakrayal** [comment]
@Saduf2019 ,
I was able to reproduce the issue in tf [v2.5](https://colab.research.google.com/gist/tilakrayal/e8b505e356c58733bc9f2c2d90daeaa5/2-5model_evaluate.ipynb), [v2.6](https://colab.research.google.com/gist/tilakrayal/a961871712b2465b4bb6c5b9031b0900/2-6-model_evaluate.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/1ea6aa835c043ffdf216d96c4c0961d9/nightly-model_evaluate.ipynb).Please find the gist of it here.

**Comment by sachinprasadhs** [comment]
For each batch during training, Keras will compute this metric and keep track of its mean since the beginning of the epoch. This is exactly what you are trying to calculate manually.
But this is not the case always.Instead of batch wise accuracy or it keeps track of the all True positives and False positives and returns the overall precision, which is called `streaming metric `or `stateful metric` as it is updated batch after batch.
Consider the below simple example.
```
>>> precision = keras.metrics.Precision()
>>> precision([0, 1, 1, 1, 0, 1, 0, 1], [1, 1, 0, 1, 0, 1, 0, 1])
<tf.Tensor: id=581729, shape=(), dtype=float32, numpy=0.8>
>>> precision([0, 1, 0, 0, 1, 0, 1, 1], [1, 0, 1, 1, 0, 0, 0, 0])
<tf.Tensor: id=581780, shape=(), dtype=float32, numpy=0.5>
```
In the above example by passing predictions and labels of first batch and second batch, you can see the it returns 0.8 for first batch and 0.5 for second batch, in second batch instead of returning precision for only second batch it actually returns the overall precision.
